// allure-js-commons/dist/src/constructors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const model_1 = require("./model");
const uuid_1 = require("uuid");
function testResultContainer() {
    return {
        uuid: uuid_1.v4(),
        children: [],
        befores: [],
        afters: []
    };
}
exports.testResultContainer = testResultContainer;
function fixtureResult() {
    return {
        status: model_1.Status.BROKEN,
        statusDetails: {},
        stage: model_1.Stage.PENDING,
        steps: [],
        attachments: [],
        parameters: []
    };
}
exports.fixtureResult = fixtureResult;
function stepResult() {
    return {
        status: undefined,
        statusDetails: {},
        stage: model_1.Stage.PENDING,
        steps: [],
        attachments: [],
        parameters: []
    };
}
exports.stepResult = stepResult;
function testResult() {
    return {
        uuid: uuid_1.v4(),
        historyId: uuid_1.v4(),
        status: undefined,
        statusDetails: {},
        stage: model_1.Stage.PENDING,
        steps: [],
        attachments: [],
        parameters: [],
        labels: [],
        links: []
    };
}
exports.testResult = testResult;
