// cucumber/lib/formatter/step_definition_snippet_builder/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _cucumberExpressions = require('cucumber-expressions');

var _helpers = require('../helpers');

var _step_arguments = require('../../step_arguments');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StepDefinitionSnippetBuilder = function () {
  function StepDefinitionSnippetBuilder(_ref) {
    var snippetSyntax = _ref.snippetSyntax,
        parameterTypeRegistry = _ref.parameterTypeRegistry;
    (0, _classCallCheck3.default)(this, StepDefinitionSnippetBuilder);

    this.snippetSyntax = snippetSyntax;
    this.cucumberExpressionGenerator = new _cucumberExpressions.CucumberExpressionGenerator(parameterTypeRegistry);
  }

  (0, _createClass3.default)(StepDefinitionSnippetBuilder, [{
    key: 'build',
    value: function build(_ref2) {
      var keywordType = _ref2.keywordType,
          pickleStep = _ref2.pickleStep;

      var comment = 'Write code here that turns the phrase above into concrete actions';
      var functionName = this.getFunctionName(keywordType);
      var generatedExpressions = this.cucumberExpressionGenerator.generateExpressions(pickleStep.text, true);
      var stepParameterNames = this.getStepParameterNames(pickleStep);
      return this.snippetSyntax.build({
        comment: comment,
        functionName: functionName,
        generatedExpressions: generatedExpressions,
        stepParameterNames: stepParameterNames
      });
    }
  }, {
    key: 'getFunctionName',
    value: function getFunctionName(keywordType) {
      switch (keywordType) {
        case _helpers.KeywordType.EVENT:
          return 'When';
        case _helpers.KeywordType.OUTCOME:
          return 'Then';
        case _helpers.KeywordType.PRECONDITION:
          return 'Given';
      }
    }
  }, {
    key: 'getStepParameterNames',
    value: function getStepParameterNames(step) {
      var iterator = (0, _step_arguments.buildStepArgumentIterator)({
        dataTable: function dataTable() {
          return 'dataTable';
        },
        docString: function docString() {
          return 'docString';
        }
      });
      return step.arguments.map(iterator);
    }
  }]);
  return StepDefinitionSnippetBuilder;
}();

exports.default = StepDefinitionSnippetBuilder;
