// allure-js-commons/dist/src/writers/InMemoryAllureWriter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class InMemoryAllureWriter {
    constructor() {
        this.groups = [];
        this.tests = [];
        this.attachments = {};
    }
    writeGroup(result) {
        this.groups.push(result);
    }
    writeResult(result) {
        this.tests.push(result);
    }
    writeAttachment(name, content) {
        this.attachments[name] = content;
    }
    writeCategoriesDefinitions(categories) {
        if (this.categories) {
            console.warn("overwriting existing categories");
        }
        this.categories = categories;
    }
    writeEnvironmentInfo(envInfo) {
        if (this.envInfo) {
            console.warn("overwriting existing environment info");
        }
        this.envInfo = envInfo;
    }
    reset() {
        this.groups = [];
        this.tests = [];
        this.attachments = {};
    }
    getMaybeTestByName(name) {
        return this.tests.find(t => t.name === name);
    }
    getTestByName(name) {
        const res = this.getMaybeTestByName(name);
        if (res === undefined)
            throw new Error(`Test not found: ${name}`);
        return res;
    }
}
exports.InMemoryAllureWriter = InMemoryAllureWriter;
