// cucumber/lib/formatter/progress_bar_formatter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _helpers = require('./helpers');

var _ = require('./');

var _2 = _interopRequireDefault(_);

var _progress = require('progress');

var _progress2 = _interopRequireDefault(_progress);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Inspired by https://github.com/thekompanee/fuubar and https://github.com/martinciu/fuubar-cucumber
var ProgressBarFormatter = function (_Formatter) {
  (0, _inherits3.default)(ProgressBarFormatter, _Formatter);

  function ProgressBarFormatter(options) {
    (0, _classCallCheck3.default)(this, ProgressBarFormatter);

    var _this = (0, _possibleConstructorReturn3.default)(this, (ProgressBarFormatter.__proto__ || Object.getPrototypeOf(ProgressBarFormatter)).call(this, options));

    options.eventBroadcaster.on('pickle-accepted', _this.incrementStepCount.bind(_this)).once('test-case-started', _this.initializeProgressBar.bind(_this)).on('test-step-finished', _this.logProgress.bind(_this)).on('test-case-finished', _this.logErrorIfNeeded.bind(_this)).on('test-run-finished', _this.logSummary.bind(_this));
    _this.numberOfSteps = 0;
    _this.issueCount = 0;
    return _this;
  }

  (0, _createClass3.default)(ProgressBarFormatter, [{
    key: 'incrementStepCount',
    value: function incrementStepCount(_ref) {
      var pickle = _ref.pickle;

      this.numberOfSteps += pickle.steps.length;
    }
  }, {
    key: 'initializeProgressBar',
    value: function initializeProgressBar() {
      this.progressBar = new _progress2.default(':current/:total steps [:bar] ', {
        clear: true,
        incomplete: ' ',
        stream: this.stream,
        total: this.numberOfSteps,
        width: this.stream.columns || 80
      });
    }
  }, {
    key: 'logProgress',
    value: function logProgress(_ref2) {
      var index = _ref2.index,
          sourceLocation = _ref2.testCase.sourceLocation;

      var _eventDataCollector$g = this.eventDataCollector.getTestCaseData(sourceLocation),
          testCase = _eventDataCollector$g.testCase;

      if (testCase.steps[index].sourceLocation) {
        this.progressBar.tick();
      }
    }
  }, {
    key: 'logErrorIfNeeded',
    value: function logErrorIfNeeded(_ref3) {
      var sourceLocation = _ref3.sourceLocation,
          result = _ref3.result;

      if ((0, _helpers.isIssue)(result.status)) {
        this.issueCount += 1;

        var _eventDataCollector$g2 = this.eventDataCollector.getTestCaseData(sourceLocation),
            gherkinDocument = _eventDataCollector$g2.gherkinDocument,
            pickle = _eventDataCollector$g2.pickle,
            testCase = _eventDataCollector$g2.testCase;

        this.progressBar.interrupt((0, _helpers.formatIssue)({
          colorFns: this.colorFns,
          gherkinDocument: gherkinDocument,
          number: this.issueCount,
          pickle: pickle,
          snippetBuilder: this.snippetBuilder,
          testCase: testCase
        }));
      }
    }
  }, {
    key: 'logSummary',
    value: function logSummary(testRun) {
      this.log((0, _helpers.formatSummary)({
        colorFns: this.colorFns,
        testCaseMap: this.eventDataCollector.testCaseMap,
        testRun: testRun
      }));
    }
  }]);
  return ProgressBarFormatter;
}(_2.default);

exports.default = ProgressBarFormatter;
