// cucumber/lib/formatter/snippets_formatter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _ = require('./');

var _2 = _interopRequireDefault(_);

var _status = require('../status');

var _status2 = _interopRequireDefault(_status);

var _helpers = require('./helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SnippetsFormatter = function (_Formatter) {
  (0, _inherits3.default)(SnippetsFormatter, _Formatter);

  function SnippetsFormatter(options) {
    (0, _classCallCheck3.default)(this, SnippetsFormatter);

    var _this = (0, _possibleConstructorReturn3.default)(this, (SnippetsFormatter.__proto__ || Object.getPrototypeOf(SnippetsFormatter)).call(this, options));

    options.eventBroadcaster.on('test-step-finished', _this.logUndefinedTestStepSnippet.bind(_this));
    return _this;
  }

  (0, _createClass3.default)(SnippetsFormatter, [{
    key: 'logUndefinedTestStepSnippet',
    value: function logUndefinedTestStepSnippet(_ref) {
      var sourceLocation = _ref.testCase.sourceLocation,
          index = _ref.index,
          result = _ref.result;

      if (result.status === _status2.default.UNDEFINED) {
        var _eventDataCollector$g = this.eventDataCollector.getTestCaseData(sourceLocation),
            gherkinDocument = _eventDataCollector$g.gherkinDocument,
            testCase = _eventDataCollector$g.testCase;

        var _eventDataCollector$g2 = this.eventDataCollector.getTestStepData({ testCase: testCase, index: index }),
            pickleStep = _eventDataCollector$g2.pickleStep,
            gherkinKeyword = _eventDataCollector$g2.gherkinKeyword;

        var previousKeywordType = this.getPreviousKeywordType({
          gherkinDocument: gherkinDocument,
          testCase: testCase,
          index: index
        });
        var keywordType = (0, _helpers.getStepKeywordType)({
          keyword: gherkinKeyword,
          language: gherkinDocument.feature.language,
          previousKeywordType: previousKeywordType
        });
        var snippet = this.snippetBuilder.build({ keywordType: keywordType, pickleStep: pickleStep });
        this.log(snippet + '\n\n');
      }
    }
  }, {
    key: 'getPreviousKeywordType',
    value: function getPreviousKeywordType(_ref2) {
      var gherkinDocument = _ref2.gherkinDocument,
          testCase = _ref2.testCase,
          index = _ref2.index;

      var previousKeywordType = _helpers.KeywordType.PRECONDITION;
      for (var i = 0; i < index; i += 1) {
        var _eventDataCollector$g3 = this.eventDataCollector.getTestStepData({
          testCase: testCase,
          index: i
        }),
            gherkinKeyword = _eventDataCollector$g3.gherkinKeyword;

        previousKeywordType = (0, _helpers.getStepKeywordType)({
          keyword: gherkinKeyword,
          language: gherkinDocument.feature.language,
          previousKeywordType: previousKeywordType
        });
      }
      return previousKeywordType;
    }
  }]);
  return SnippetsFormatter;
}(_2.default);

exports.default = SnippetsFormatter;
