// cucumber/lib/runtime/attachment_manager/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _isStream = require('is-stream');

var _isStream2 = _interopRequireDefault(_isStream);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AttachmentManager = function () {
  function AttachmentManager(onAttachment) {
    (0, _classCallCheck3.default)(this, AttachmentManager);

    this.onAttachment = onAttachment;
  }

  (0, _createClass3.default)(AttachmentManager, [{
    key: 'create',
    value: function create(data, mediaType, callback) {
      if (Buffer.isBuffer(data)) {
        if (!mediaType) {
          throw Error('Buffer attachments must specify a media type');
        }
        this.createBufferAttachment(data, mediaType);
      } else if (_isStream2.default.readable(data)) {
        if (!mediaType) {
          throw Error('Stream attachments must specify a media type');
        }
        return this.createStreamAttachment(data, mediaType, callback);
      } else if (typeof data === 'string') {
        if (!mediaType) {
          mediaType = 'text/plain';
        }
        this.createStringAttachment(data, { type: mediaType });
      } else {
        throw Error('Invalid attachment data: must be a buffer, readable stream, or string');
      }
    }
  }, {
    key: 'createBufferAttachment',
    value: function createBufferAttachment(data, mediaType) {
      this.createStringAttachment(data.toString('base64'), {
        encoding: 'base64',
        type: mediaType
      });
    }
  }, {
    key: 'createStreamAttachment',
    value: function createStreamAttachment(data, mediaType, callback) {
      var _this = this;

      var promise = new _bluebird2.default(function (resolve, reject) {
        var buffers = [];
        data.on('data', function (chunk) {
          buffers.push(chunk);
        });
        data.on('end', function () {
          _this.createBufferAttachment(Buffer.concat(buffers), mediaType);
          resolve();
        });
        data.on('error', reject);
      });
      if (callback) {
        promise.then(callback, callback);
      } else {
        return promise;
      }
    }
  }, {
    key: 'createStringAttachment',
    value: function createStringAttachment(data, media) {
      this.onAttachment({ data: data, media: media });
    }
  }]);
  return AttachmentManager;
}();

exports.default = AttachmentManager;
